File: tasks/native-runtime/budget_ledger_parity_probe.py
"""Budget-ledger parity probe, Python side.

Covers the behavioural half of `budget_manager.py` lines 183-371: the spend view a row becomes,
the row `record_spend` writes, the four threshold checks, `should_downgrade`,
`record_request_spend` and `budget_status`.

Two deliberate stubs, matching what the Rust side injects:

- **the connection.** Group B patches `connect_db` to a real in-memory database and points
  `BUDGET_DB` at a file that exists, so the oracle's own SQL runs here. On the Rust side the
  same upsert semantics are simulated by a map, because the SQL statements and the connection
  are **not part of this slice** — they belong to the store. What the two sides therefore
  compare is the shaping around the SQL (floors, rounding, the timestamp) and the read-back
  view, not the statements themselves.
- **`daily_spend`**, in group A, so the threshold and status logic can be driven from fixed
  spend views instead of from a database.

Usage::

    python tasks/native-runtime/budget_ledger_parity_probe.py > python.json
    cd rust && cargo run -p deepseek-policy --example budget_ledger_parity_probe > ../rust.json
"""

from __future__ import annotations

import json
import sqlite3
import sys
import tempfile
from pathlib import Path
from typing import Any

REPO = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(REPO))

from deepseek_infra.infra.gateway import budget_manager as bm  # noqa: E402

SPEND_CASES: list[dict[str, Any]] = [
    {"totalTokens": 0, "costUsd": 0.0, "searchCalls": 0, "toolCalls": 0},
    {"totalTokens": 100, "costUsd": 0.5, "searchCalls": 2, "toolCalls": 3},
    {"totalTokens": 1_000, "costUsd": 1.5, "searchCalls": 3, "toolCalls": 4},
    {"totalTokens": 999, "costUsd": 1.499_999, "searchCalls": 2, "toolCalls": 3},
]

POLICY_CASES: list[dict[str, Any]] = [
    {"max_total_tokens": 0, "max_agent_tokens": 0, "max_search_calls": 0, "max_tool_calls": 0, "max_estimated_cost_usd": 0.0, "policy": "none"},
    {"max_total_tokens": 1_000, "max_search_calls": 3, "max_tool_calls": 4, "max_estimated_cost_usd": 1.5, "policy": "none"},
    {"max_total_tokens": 100, "max_search_calls": 0, "max_tool_calls": 0, "max_estimated_cost_usd": 0.0, "policy": "none"},
    {"max_total_tokens": 1, "max_search_calls": 1, "max_tool_calls": 1, "max_estimated_cost_usd": 0.01, "policy": "downgrade_to_flash_when_exceeded"},
    {"max_total_tokens": -5, "max_search_calls": -1, "max_tool_calls": -1, "max_estimated_cost_usd": -1.0, "policy": "none"},
    {"max_total_tokens": 0, "max_search_calls": 0, "max_tool_calls": 0, "max_estimated_cost_usd": 0.0, "policy": "downgrade_to_flash_when_exceeded"},
]

RECORD_CASES: list[dict[str, Any]] = [
    {"scope": "global", "prompt_tokens": 10, "completion_tokens": 20, "cost_usd": 4.9e-05, "model_calls": 1},
    {"scope": "global", "prompt_tokens": 5, "completion_tokens": 0, "cost_usd": 0.000_075_5, "model_calls": 1, "tool_calls": 2},
    {"scope": "项目:1", "prompt_tokens": -5, "completion_tokens": 3.7, "cost_usd": -1.0, "model_calls": -2, "search_calls": "2"},
    {"scope": "", "prompt_tokens": " 7 ", "completion_tokens": "abc", "cost_usd": "0.5", "model_calls": 0},
]

REQUEST_SPEND_CASES: list[Any] = [
    ({}, None, {}, 0, 0),
    ({"memoryScope": "project:abc"}, "deepseek-v4-pro", {"prompt_tokens": 100, "completion_tokens": 50}, 2, 1),
    ({}, "deepseek-v4-flash", {"promptTokens": "30"}, -3, 0),
    ({"projectId": "p1"}, "unknown", "not-a-dict", 1, 1),
]


def policy(case: dict[str, Any]) -> Any:
    return bm.BudgetPolicy(
        max_total_tokens=case["max_total_tokens"],
        max_agent_tokens=case.get("max_agent_tokens", 0),
        max_search_calls=case["max_search_calls"],
        max_tool_calls=case["max_tool_calls"],
        max_estimated_cost_usd=case["max_estimated_cost_usd"],
        policy=case["policy"],
    )


def spend_view(case: dict[str, Any]) -> dict[str, Any]:
    return {
        "scope": "global",
        "day": "2026-09-18",
        "promptTokens": case["totalTokens"],
        "completionTokens": 0,
        "totalTokens": case["totalTokens"],
        "costUsd": case["costUsd"],
        "modelCalls": 1,
        "searchCalls": case["searchCalls"],
        "toolCalls": case["toolCalls"],
    }


def main() -> int:
    out: dict[str, Any] = {}

    # --- A: thresholds, the downgrade decision and the status envelope -----------------
    daily_calls = {"count": 0}
    failures = {"connect": False}

    def stubbed_daily_spend(scope: str = "global", day: str | None = None) -> dict[str, Any]:
        daily_calls["count"] += 1
        return CURRENT_SPEND[0]

    CURRENT_SPEND: list[dict[str, Any]] = [spend_view(SPEND_CASES[0])]
    real_daily_spend = bm.daily_spend
    real_budget_db = bm.BUDGET_DB
    # `budget_status` renders the path, so both sides use the same relative one, and a file
    # has to exist for the read path to be taken (the connection itself is stubbed).
    bm.BUDGET_DB = Path("probe-budget.db")
    probe_db = bm.BUDGET_DB
    probe_db.write_text("", encoding="utf-8")
    bm.daily_spend = stubbed_daily_spend
    try:
        for spend_index, spend_case in enumerate(SPEND_CASES):
            CURRENT_SPEND[0] = spend_view(spend_case)
            for policy_index, policy_case in enumerate(POLICY_CASES):
                over = bm.over_daily_budget("global", policy(policy_case))
                out[f"over::{spend_index}::{policy_index}"] = over
                out[f"downgrade::{spend_index}::{policy_index}"] = bm.should_downgrade(
                    "global", policy(policy_case)
                )

        for scope_index, scope in enumerate(["global", "project:abc", ""]):
            daily_calls["count"] = 0
            status = bm.budget_status(scope)
            out[f"status::{scope_index}"] = status
            # `today` and `overBudget` each read the ledger: two reads, not one.
            out[f"status-reads::{scope_index}"] = daily_calls["count"]
    finally:
        bm.daily_spend = real_daily_spend

    # --- B: the row, the upsert and the read-back -------------------------------------
    saved = (bm.connect_db, bm.BUDGET_DIR, bm.BUDGET_TRACKING_ENABLED, bm.today)
    temp_dir = Path(tempfile.mkdtemp(prefix="budget-probe-"))
    try:
        connection = sqlite3.connect(":memory:")
        connection.row_factory = sqlite3.Row
        bm.connect_db = lambda: connection
        bm.BUDGET_DIR = temp_dir
        bm.BUDGET_TRACKING_ENABLED = True
        bm.today = lambda: "2026-09-18"

        out["read::missing-row"] = bm.daily_spend("global")
        out["read::missing-scope"] = bm.daily_spend("project:never")

        for index, case in enumerate(RECORD_CASES):
            bm.record_spend(
                case["scope"],
                prompt_tokens=case["prompt_tokens"],
                completion_tokens=case["completion_tokens"],
                cost_usd=case["cost_usd"],
                model_calls=case["model_calls"],
                search_calls=case.get("search_calls", 0),
                tool_calls=case.get("tool_calls", 0),
            )
            out[f"record::{index}"] = bm.daily_spend(case["scope"])

        # The same row again: an upsert accumulates rather than replacing.
        bm.record_spend("global", prompt_tokens=1, completion_tokens=1, cost_usd=1e-06, model_calls=1)
        out["record::accumulated"] = bm.daily_spend("global")
        out["record::other-day"] = bm.daily_spend("global", "2026-01-01")

        # Recording is gated; reading still works while it is off.
        bm.BUDGET_TRACKING_ENABLED = False
        bm.record_spend("gated", prompt_tokens=9, completion_tokens=9, cost_usd=1.0, model_calls=1)
        out["gated::write"] = bm.daily_spend("gated")
        bm.BUDGET_TRACKING_ENABLED = True

        # A database file that is not there is the empty view, not an error.
        bm.BUDGET_DB = temp_dir / "absent.db"
        out["absent::read"] = bm.daily_spend("global")

        # A read that blows up is stored, not raised: the empty view plus `lastError`.
        bm.BUDGET_DB = probe_db
        bm.connect_db = lambda: (_ for _ in ()).throw(sqlite3.OperationalError("database is locked"))
        failed = bm.daily_spend("global")
        out["failed::read"] = failed
        out["failed::status"] = bm.budget_status("global")
        out["failed::error"] = bm._last_error
        bm.connect_db = lambda: connection

        for index, (payload, model, usage, tool_calls, search_calls) in enumerate(REQUEST_SPEND_CASES):
            view = bm.record_request_spend(
                payload, model, usage, tool_calls=tool_calls, search_calls=search_calls
            )
            out[f"request-spend::{index}"] = view
        out["request-spend::scope-total"] = bm.daily_spend("global")
        out["request-spend::project-total"] = bm.daily_spend("project:abc")
        out["request-spend::p1-total"] = bm.daily_spend("project:p1")
        out["request-spend::free-total"] = bm.daily_spend("global", "2026-01-01")
    finally:
        bm.connect_db, bm.BUDGET_DIR, bm.BUDGET_TRACKING_ENABLED, bm.today = saved
        bm.BUDGET_DB = real_budget_db
        probe_db.unlink(missing_ok=True)

    json.dump(out, sys.stdout, ensure_ascii=False, indent=2, sort_keys=True)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: rust/crates/deepseek-policy/examples/budget_ledger_parity_probe.rs
//! Budget-ledger parity probe, Rust side.
//!
//! Replays the same corpus as `tasks/native-runtime/budget_ledger_parity_probe.py`.
//!
//! The Python probe runs the oracle's own SQL against an in-memory database. **This side has
//! no SQL** — the statements and the connection are the store's slice — so the upsert
//! semantics are simulated by a map whose accumulate/overwrite behaviour mirrors the
//! `ON CONFLICT ... DO UPDATE SET` clause. What the two sides therefore compare is the
//! shaping around the SQL (the `max(0, int(...))` floors, the six-decimal cost, the
//! timestamp) and the view read back, plus the threshold, downgrade and status logic.
//!
//! `last_error` is threaded as a local, mirroring the oracle's module-level `_last_error`:
//! `budget_status` reads it back out, so whichever read failed most recently wins.
//!
//! Usage::
//!
//!     python tasks/native-runtime/budget_ledger_parity_probe.py > python.json
//!     cd rust && cargo run -p deepseek-policy --example budget_ledger_parity_probe > ../rust.json

use std::cell::RefCell;

use serde_json::{Map, Value, json};

use deepseek_policy::budget_ledger::{
    LedgerDeps, budget_status, daily_spend, over_daily_budget, record_request_spend, record_spend,
    should_downgrade,
};
use deepseek_policy::budget_manager::{BudgetPolicy, BudgetSettings};
use deepseek_policy::python_json::OrderedJson;

const DAY: &str = "2026-09-18";
const NOW_ISO: &str = "2026-09-18T00:00:00Z";

fn spend_cases() -> Vec<Value> {
    vec![
        json!({"totalTokens": 0, "costUsd": 0.0, "searchCalls": 0, "toolCalls": 0}),
        json!({"totalTokens": 100, "costUsd": 0.5, "searchCalls": 2, "toolCalls": 3}),
        json!({"totalTokens": 1_000, "costUsd": 1.5, "searchCalls": 3, "toolCalls": 4}),
        json!({"totalTokens": 999, "costUsd": 1.499_999, "searchCalls": 2, "toolCalls": 3}),
    ]
}

fn policy_cases() -> Vec<Value> {
    vec![
        json!({"max_total_tokens": 0, "max_agent_tokens": 0, "max_search_calls": 0, "max_tool_calls": 0, "max_estimated_cost_usd": 0.0, "policy": "none"}),
        json!({"max_total_tokens": 1_000, "max_search_calls": 3, "max_tool_calls": 4, "max_estimated_cost_usd": 1.5, "policy": "none"}),
        json!({"max_total_tokens": 100, "max_search_calls": 0, "max_tool_calls": 0, "max_estimated_cost_usd": 0.0, "policy": "none"}),
        json!({"max_total_tokens": 1, "max_search_calls": 1, "max_tool_calls": 1, "max_estimated_cost_usd": 0.01, "policy": "downgrade_to_flash_when_exceeded"}),
        json!({"max_total_tokens": -5, "max_search_calls": -1, "max_tool_calls": -1, "max_estimated_cost_usd": -1.0, "policy": "none"}),
        json!({"max_total_tokens": 0, "max_search_calls": 0, "max_tool_calls": 0, "max_estimated_cost_usd": 0.0, "policy": "downgrade_to_flash_when_exceeded"}),
    ]
}

fn record_cases() -> Vec<(&'static str, Value)> {
    vec![
        (
            "global",
            json!({"prompt_tokens": 10, "completion_tokens": 20, "cost_usd": 4.9e-05, "model_calls": 1}),
        ),
        (
            "global",
            json!({"prompt_tokens": 5, "completion_tokens": 0, "cost_usd": 0.000_075_5, "model_calls": 1, "tool_calls": 2}),
        ),
        (
            "项目:1",
            json!({"prompt_tokens": -5, "completion_tokens": 3.7, "cost_usd": -1.0, "model_calls": -2, "search_calls": "2"}),
        ),
        (
            "",
            json!({"prompt_tokens": " 7 ", "completion_tokens": "abc", "cost_usd": "0.5", "model_calls": 0}),
        ),
    ]
}

fn request_spend_cases() -> Vec<(Value, Option<&'static str>, Value, i64, i64)> {
    vec![
        (json!({}), None, json!({}), 0, 0),
        (
            json!({"memoryScope": "project:abc"}),
            Some("deepseek-v4-pro"),
            json!({"prompt_tokens": 100, "completion_tokens": 50}),
            2,
            1,
        ),
        (
            json!({}),
            Some("deepseek-v4-flash"),
            json!({"promptTokens": "30"}),
            -3,
            0,
        ),
        (
            json!({"projectId": "p1"}),
            Some("unknown"),
            json!("not-a-dict"),
            1,
            1,
        ),
    ]
}

fn policy(case: &Value) -> BudgetPolicy {
    BudgetPolicy {
        max_total_tokens: case["max_total_tokens"].as_i64().unwrap_or(0),
        max_agent_tokens: case["max_agent_tokens"].as_i64().unwrap_or(0),
        max_search_calls: case["max_search_calls"].as_i64().unwrap_or(0),
        max_tool_calls: case["max_tool_calls"].as_i64().unwrap_or(0),
        max_estimated_cost_usd: case["max_estimated_cost_usd"].as_f64().unwrap_or(0.0),
        policy: case["policy"].as_str().unwrap_or("none").to_string(),
    }
}

/// The row whose spend view equals the Python probe's stubbed view.
fn row_for(case: &Value) -> Value {
    json!({
        "scope": "global",
        "day": DAY,
        "prompt_tokens": case["totalTokens"],
        "completion_tokens": 0,
        "cost_usd": case["costUsd"],
        "model_calls": 1,
        "search_calls": case["searchCalls"],
        "tool_calls": case["toolCalls"],
    })
}

fn main() {
    let mut out: Map<String, Value> = Map::new();
    let settings = BudgetSettings::default();
    let mut last_error = String::new();

    // --- A: thresholds, the downgrade decision and the status envelope -----------------
    let spend_cases = spend_cases();
    let policy_cases = policy_cases();
    let current: RefCell<Value> = RefCell::new(row_for(&spend_cases[0]));
    let read_current = |_scope: &str, _day: &str| -> Result<Option<Value>, String> {
        Ok(Some(current.borrow().clone()))
    };
    let unreachable_write = |_row: &Value| -> Result<(), String> { Ok(()) };
    let reads: RefCell<i64> = RefCell::new(0);
    let counting = |scope: &str, day: &str| -> Result<Option<Value>, String> {
        *reads.borrow_mut() += 1;
        read_current(scope, day)
    };
    let deps = LedgerDeps {
        database_present: true,
        database_path: "probe-budget.db".to_string(),
        day: DAY.to_string(),
        now_iso: NOW_ISO.to_string(),
        read_spend_row: &counting,
        write_spend_row: &unreachable_write,
    };

    for (spend_index, spend_case) in spend_cases.iter().enumerate() {
        *current.borrow_mut() = row_for(spend_case);
        for (policy_index, policy_case) in policy_cases.iter().enumerate() {
            let policy = policy(policy_case);
            let (spend, error) = daily_spend("global", None, &deps);
            assert!(error.is_none());
            let over = over_daily_budget(&spend, &policy);
            out.insert(format!("over::{spend_index}::{policy_index}"), over.clone());
            out.insert(
                format!("downgrade::{spend_index}::{policy_index}"),
                json!(should_downgrade(&policy, &over)),
            );
        }
    }

    for (scope_index, scope) in ["global", "project:abc", ""].iter().enumerate() {
        *reads.borrow_mut() = 0;
        let status = budget_status(scope, &json!({}), &settings, &deps, &last_error);
        out.insert(format!("status::{scope_index}"), status);
        // `today` and `overBudget` each read the ledger: two reads, not one.
        out.insert(
            format!("status-reads::{scope_index}"),
            json!(*reads.borrow()),
        );
    }

    // --- B: the row, the upsert and the read-back -------------------------------------
    let table: RefCell<Map<String, Value>> = RefCell::new(Map::new());
    let write = |row: &Value| -> Result<(), String> {
        let key = format!(
            "{}|{}",
            row.get("scope").and_then(Value::as_str).unwrap_or_default(),
            row.get("day").and_then(Value::as_str).unwrap_or_default()
        );
        let mut table = table.borrow_mut();
        match table.get_mut(&key) {
            None => {
                table.insert(key, row.clone());
            }
            Some(existing) => {
                let existing = existing.as_object_mut().expect("a row is an object");
                for field in [
                    "prompt_tokens",
                    "completion_tokens",
                    "model_calls",
                    "search_calls",
                    "tool_calls",
                ] {
                    let current = existing.get(field).and_then(Value::as_i64).unwrap_or(0);
                    let incoming = row.get(field).and_then(Value::as_i64).unwrap_or(0);
                    existing.insert(field.to_string(), json!(current + incoming));
                }
                let current = existing
                    .get("cost_usd")
                    .and_then(Value::as_f64)
                    .unwrap_or(0.0);
                let incoming = row.get("cost_usd").and_then(Value::as_f64).unwrap_or(0.0);
                existing.insert("cost_usd".to_string(), json!(current + incoming));
                existing.insert(
                    "updated_at".to_string(),
                    row.get("updated_at").cloned().unwrap_or(Value::Null),
                );
            }
        }
        Ok(())
    };
    let read = |scope: &str, day: &str| -> Result<Option<Value>, String> {
        Ok(table.borrow().get(&format!("{scope}|{day}")).cloned())
    };
    let locked = |_scope: &str, _day: &str| -> Result<Option<Value>, String> {
        Err("database is locked".to_string())
    };

    let mut live = LedgerDeps {
        database_present: true,
        database_path: "probe-budget.db".to_string(),
        day: DAY.to_string(),
        now_iso: NOW_ISO.to_string(),
        read_spend_row: &read,
        write_spend_row: &write,
    };

    let (missing, error) = daily_spend("global", None, &live);
    assert!(error.is_none());
    out.insert("read::missing-row".to_string(), missing);
    let (missing_scope, _) = daily_spend("project:never", None, &live);
    out.insert("read::missing-scope".to_string(), missing_scope);

    for (index, (scope, fields)) in record_cases().iter().enumerate() {
        if let Some(error) = record_spend(&json!(scope), fields, &settings, &live) {
            last_error = error;
        }
        let (spend, read_error) = daily_spend(scope, None, &live);
        if let Some(error) = read_error {
            last_error = error;
        }
        out.insert(format!("record::{index}"), spend);
    }

    if let Some(error) = record_spend(
        &json!("global"),
        &json!({"prompt_tokens": 1, "completion_tokens": 1, "cost_usd": 1e-06, "model_calls": 1}),
        &settings,
        &live,
    ) {
        last_error = error;
    }
    let (accumulated, _) = daily_spend("global", None, &live);
    out.insert("record::accumulated".to_string(), accumulated);
    let (other_day, _) = daily_spend("global", Some("2026-01-01"), &live);
    out.insert("record::other-day".to_string(), other_day);

    let gated_settings = BudgetSettings {
        tracking_enabled: false,
        ..BudgetSettings::default()
    };
    if let Some(error) = record_spend(
        &json!("gated"),
        &json!({"prompt_tokens": 9, "completion_tokens": 9, "cost_usd": 1.0, "model_calls": 1}),
        &gated_settings,
        &live,
    ) {
        last_error = error;
    }
    let (gated, _) = daily_spend("gated", None, &live);
    out.insert("gated::write".to_string(), gated);

    live.database_present = false;
    let (absent, error) = daily_spend("global", None, &live);
    assert!(error.is_none());
    out.insert("absent::read".to_string(), absent);
    live.database_present = true;

    let failing = LedgerDeps {
        database_present: true,
        database_path: "probe-budget.db".to_string(),
        day: DAY.to_string(),
        now_iso: NOW_ISO.to_string(),
        read_spend_row: &locked,
        write_spend_row: &unreachable_write,
    };
    let (failed, read_error) = daily_spend("global", None, &failing);
    if let Some(error) = read_error {
        last_error = error;
    }
    out.insert("failed::read".to_string(), failed);
    out.insert(
        "failed::status".to_string(),
        budget_status("global", &json!({}), &settings, &failing, &last_error),
    );
    out.insert("failed::error".to_string(), json!(last_error));

    for (index, (payload, model, usage, tool_calls, search_calls)) in
        request_spend_cases().iter().enumerate()
    {
        // Nothing reads the ledger's last error after this point, on either side.
        let (view, _error) = record_request_spend(
            payload,
            *model,
            usage,
            *tool_calls,
            *search_calls,
            &settings,
            &live,
        );
        out.insert(format!("request-spend::{index}"), view);
    }
    for (key, scope, day) in [
        ("request-spend::scope-total", "global", None),
        ("request-spend::project-total", "project:abc", None),
        ("request-spend::p1-total", "project:p1", None),
        ("request-spend::free-total", "global", Some("2026-01-01")),
    ] {
        let (spend, _) = daily_spend(scope, day, &live);
        out.insert(key.to_string(), spend);
    }

    let rendered = OrderedJson::from_value_with_order(&Value::Object(out), &[]).render_indent_2();
    println!("{rendered}");
}
